### DISCUSSION TITLE
How to add variables using golang ?

### DISCUSSION THREAD

**Discussion by apodgorbunschih** [terrastruct/d2#1920] [Q&A]
I want to use d2lang in a golang project to draw some diagrams. Is it possible to set `variables`  dynamically ?

```
	// From one.go
	ruler, _ := textmeasure.NewRuler()
	layoutResolver := func(engine string) (d2graph.LayoutGraph, error) {
		return d2dagrelayout.DefaultLayout, nil
	}
	compileOpts := &d2lib.CompileOptions{
		LayoutResolver: layoutResolver,
		Ruler:          ruler,
	}
	_, graph, _ := d2lib.Compile(context.Background(), "x -> y", compileOpts, nil)

	// Create a shape with the ID, "meow"
	graph, _, _ = d2oracle.Create(graph, nil, "meow")
	// Style the shape green
	color := "green"
	server := "us-east-1"
	graph, _ = d2oracle.Set(graph, nil, "meow.style.fill", nil, &color)
        graph, _ = d2oracle.Set(graph, nil, "meow.vars.server", nil, &server) // <----------------------Something like that ?
	// Create a shape with the ID, "cat"
	graph, _, _ = d2oracle.Create(graph, nil, "cat")
	// Move the shape "meow" inside the container "cat"
	graph, _ = d2oracle.Move(graph, nil, "meow", "cat.meow", false)
	// Prints formatted D2 script
	fmt.Print(d2format.Format(graph.AST))
```

**Comment by alixander** [comment]
@apodgorbunschih That functionality isn't in the API yet. I made an issue. https://github.com/terrastruct/d2/issues/1922

Would you be open to sharing what your project is and how vars helps? Just helps to attach some motivations to issues

**Comment by apodgorbunschih** [comment]
@alixander awesome. I want to create some topology map for a infrastructure, i am querying the api and creating the nodes and edges step by step. So at some point i need to store some data for linking purposes, and i thought that `var` would be a good place. At the moment i am keeping the data in memory (maybe will move to a storage)

btw. Awesome tool. Playing around for couple of days, and its awesome
